plugins/woocommerce/tests/php/src/Blocks/Domain/Services/CheckoutFieldsFrontendTest.php

```diff
@@ -0,0 +1,281 @@
+	 *
+	 * @param string $message The message string.
+	 */
+	public function capture_success( string $message ): void {
+		$this->capture_messages( $message, 'success' );
+	}
+
+	/**
+	 * Capture error messages.
+	 *
+	 * @param string $message The message string.
+	 */
+	public function capture_error( string $message ): void {
+		$this->capture_messages( $message, 'error' );
+	}
+
+	/**
+	 * Capture notice messages.
+	 *
+	 * @param string $message The message string.
+	 */
+	public function capture_notice( string $message ): void {
+		$this->capture_messages( $message, 'notice' );
+	}
+
+	/**
+	 * Capture messages.
+	 *
+	 * @param string $message The message string.
+	 * @param string $type The message type.
+	 */
+	private function capture_messages( string $message, string $type ): void {
+		global $mocked_messages;
+		$mocked_messages[] = [
+			'message' => $message,
+			'type'    => $type,
+		];
+	}
+
+
+	/**
+	 * @testDox Save account form fields.
+	 */
+	public function test_save_account_form_fields_contact_customer_data_store_exception() {
+
+		$hash = uniqid();
+
+		$mock_data_store = function () use ( $hash ) {
+			throw new Exception( $hash ); // phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+		};
+
+		add_filter( 'woocommerce_customer_data_store', $mock_data_store );
+
+		global $mocked_messages;
+		$mocked_messages = [];
+
+		$this->sut->save_account_form_fields( 12345 );
+
+		remove_filter( 'woocommerce_customer_data_store', $mock_data_store );
+
+		$this->assertCount( 1, $mocked_messages );
+		$this->assertEquals( __( 'Unable to save customer additional fields: ', 'woocommerce' ) . $hash, $mocked_messages[0]['message'] );
+		$this->assertEquals( 'notice', $mocked_messages[0]['type'] );
+	}
+
+	/**
+	 * @testDox Contact additional field validation error
+	 */
+	public function test_save_account_form_fields_contact_generic_validation_error() {
+
+		woocommerce_register_additional_checkout_field(
+			array(
+				'id'                => 'mynamespace/generic_validation_error',
+				'label'             => 'Optional field with validation',
+				'location'          => 'contact',
+				'required'          => false,
+				'validate_callback' => function () {
+					return new WP_Error(
+						'generic_validation_error',
+						'Generic validation error message'
+					);
+				},
+			),
+		);
+
+		$_POST['mynamespace/generic_validation_error'] = 'value';
+		global $mocked_messages;
+		$mocked_messages = [];
+
+		$this->sut->save_account_form_fields( 1 );
+
+		$this->assertCount( 1, $mocked_messages );
+		$this->assertEquals( 'Generic validation error message', $mocked_messages[0]['message'] );
+		$this->assertEquals( 'error', $mocked_messages[0]['type'] );
+
+		$value = $this->controller->get_field_from_object( 'mynamespace/generic_validation_error', new WC_Customer( 1 ) );
+		$this->assertEquals( '', $value );
+
+		$_POST['mynamespace/generic_validation_error'] = '';
+		$mocked_messages                               = [];
```

Comment:
It might be a good idea to save this with a valid value first, then save it again with an empty value. This is so verify that going from valid -> empty works correctly. Currently we're only testing empty -> empty works, right?

Reply:
cc/ @mikejolley since this was merged to your PR and going to be closed now.